### PR TITLE
[SPARK-7758][SQL]Override more configs to avoid failure when connect to a postgre sql

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
https://issues.apache.org/jira/browse/SPARK-7758

When initializing `executionHive`, we only masks
`javax.jdo.option.ConnectionURL` to override metastore location.  However,
other properties that relates to the actual Hive metastore data source are not
masked.  For example, when using Spark SQL with a PostgreSQL backed Hive
metastore, `executionHive` actually tries to use settings read from
`hive-site.xml`, which talks about PostgreSQL, to connect to the temporary
Derby metastore, thus causes error.

To fix this, we need to mask all metastore data source properties.
Specifically, according to the code of [Hive `ObjectStore.getDataSourceProps()`
method](https://github.com/apache/hive/blob/release-0.13.1/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L288), all properties whose name mentions "jdo" and "datanucleus" must be
included.

Have tested using postgre sql as metastore, it worked fine.

### CODE REVIEW

**Comment by WangTaoTheTonic** [comment]
Jenkins, retest this please.

**Comment by WangTaoTheTonic** [comment]
From the failed test cases, it looks like not a good fix. I have no more idea as not a spark-sql expert. Anyone could offer some help?

**Comment by marmbrus** [comment]
Yeah, we definitely need the execution hive for Hive to function, but likely what we want to do is filter out any configuration that is telling it to use postgres (it should always use a dummy local derby instance)

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 1]
```diff
@@ -149,22 +149,6 @@ class HiveContext(sc: SparkContext) extends SQLContext(sc) {
...
-   * client is used for execution related tasks like registering temporary functions or ensuring
-   * that the ThreadLocal SessionState is correctly populated.  This copy of Hive is *not* used
-   * for storing peristent metadata, and only point to a dummy metastore in a temporary directory.
-   */
-  @transient
-  protected[hive] lazy val executionHive: ClientWrapper = {
-    logInfo(s"Initilizing execution hive, version $hiveExecutionVersion")
-    new ClientWrapper(
-      version = IsolatedClientLoader.hiveVersion(hiveExecutionVersion),
-      config = newTemporaryConfiguration())
```
Comment (marmbrus) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: I would try overriding more settings here.

Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: @marmbrus, i am not understand here, can you explain, why we need `executionHive`? is hive also do like this(use a local dummy metastore for temporary functions)?

**Comment by WangTaoTheTonic** [comment]
@marmbrus I overrided more configs in HiveConf (refer to https://github.com/pwendell/hive/blob/trunk/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L288) plus `datanucleus.rdbms.datastoreAdapterClassName` not included in HiveConf.

Have tested using postgre sql as metastore, it worked fine.

**Comment by yhuai** [comment]
@WangTaoTheTonic Seems `newTemporaryConfiguration` is used by `executionHive`. Can you explain the reason that your change can fix the problem of `metadataHive`?

**Comment by liancheng** [comment]
@yhuai We headed to the wrong direction at first. It's not because `executionHive` can't find proper PostgreSQL configurations. The reason of the failure is that `metadataHive` only overrides the metastore location without overriding other JDO and Datanucleus properties, thus those properties are read from `hive-site.xml` and talking about PostgreSQL. This makes `metadataHive` trying to connect to the temporary Derby metastore with PostgreSQL settings, and causes the error. What @WangTaoTheTonic did is overrides all related properties with Hive default values (gathered from `ConfVars`). That's why updating `executionHive` related code path corrects the behavior.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala (hunk 1 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: @scwf This is a Spark SQL specific design, which enables us to connect multiple versions of Hive metastore with a single code base. The `executionHive` is used for internal jobs within Spark SQL framework, while the `metadataHive` is used to talk to external Hive metastore.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 2]
```diff
@@ -507,8 +508,17 @@ private[hive] object HiveContext {
   def newTemporaryConfiguration(): Map[String, String] = {
     val tempDir = Utils.createTempDir()
     val localMetastore = new File(tempDir, "metastore").getAbsolutePath
-    Map(
-      "javax.jdo.option.ConnectionURL" -> s"jdbc:derby:;databaseName=$localMetastore;create=true")
+    val propMap: HashMap[String, String] = HashMap()
+    HiveConf.ConfVars.values().foreach { confvar  =>
+      if (confvar.varname.contains("datanucleus") || confvar.varname.contains("jdo")) {
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: Could you please add a comment to explain where this `if` condition comes from for future reference? Thanks

**Comment by liancheng** [comment]
@WangTaoTheTonic Would you mind to help updating the PR description to the following? Thanks!

> When initializing `executionHive`, we only masks `javax.jdo.option.ConnectionURL` to override metastore location.  However, other properties that relates to the actual Hive metastore data source are not masked.  For example, when using Spark SQL with a PostgreSQL backed Hive metastore, `executionHive` actually tries to use settings read from `hive-site.xml`, which talks about PostgreSQL, to connect to the temporary Derby metastore, thus causes error.
>
> To fix this, we need to mask all metastore data source properties.  Specifically, according to the code of [Hive `ObjectStore.getDataSourceProps()` method] [1], all properties whose name mentions "jdo" and "datanucleus" must be included.
>
> [1]: https://github.com/apache/hive/blob/release-0.13.1/metastore/src/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L288

**Comment by liancheng** [comment]
Forgot to say, LGTM, and thanks for fixing this!

**Comment by WangTaoTheTonic** [comment]
@yhuai I think @liancheng explains all clearly. Thanks for your comments and explanation @liancheng.

**Comment by yhuai** [comment]
@WangTaoTheTonic  Can you update your pr description per Cheng?

**Comment by marmbrus** [comment]
Thanks for fixing this!  Merged to master and 1.4.

**Pull request closed without merging**